### PR TITLE
Document javascript: URLs

### LINKED ISSUE #7143: Content suggestion: javascript pseudo-protocol identifier
## What is the new suggestion?
Document the javascript pseudo-protocol identifier `javascript:` which supposedly can be used anywhere a URL can to run javascript? Someone who has been doing web developer longer than I have can write about the history.

## Why is it important or useful?
It is supposedly a pseudo-protocol, meaning it was/is supported on older browsers? The info that exists is scattered across old blogposts and stack overflow answers. Given the sheer number of duplicate stack overflow questions, it is not a rare topic.

Probably important for XSS.

## Other supporting information

Answers are vague about history and browser support

https://stackoverflow.com/questions/2321469/when-do-i-need-to-specify-the-javascript-protocol

https://stackoverflow.com/questions/10068781/what-does-the-javascript-pseudo-protocol-actually-do?noredirect=1&lq=1

https://crisp.tweakblogs.net/blog/313/the-useless-javascript-pseudo-protocol.html from 2006

https://social.msdn.microsoft.com/Forums/ie/en-US/ac6bc491-e3d3-49eb-a010-8c36bd068016/javascript-pseudoprotocol-doesnt-work-why?forum=iewebdevelopment

### PR DESCRIPTION
**Pull request by Josh-Cena**
Fix https://github.com/mdn/content/issues/7143

### CODE REVIEW

File: files/en-us/web/uri/schemes/javascript/index.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+title: "javascript: URLs"
+slug: Web/URI/Schemes/javascript
+page-type: guide
+spec-urls: https://html.spec.whatwg.org/multipage/browsing-the-web.html#the-javascript:-url-special-case
+---
+
+{{QuickLinksWithSubpages("/en-US/docs/Web/URI")}}
+
+> [!WARNING]
+> Using `javascript:` URLs on the web is discouraged as it may lead to execution of arbitrary code, similar to the ramifications of using [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval). It may also be inaccessible because it deviates from normal link behavior.
```
Comment (hamishwillee) [inline files/en-us/web/uri/schemes/javascript/index.md:11]: What do you mean by inaccessible here?

Comment (Josh-Cena) [reply files/en-us/web/uri/schemes/javascript/index.md:11]: I means "bad in terms of accessibility". Is there a better phrase?

File: files/en-us/web/uri/schemes/javascript/index.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+`javascript:` URLs, URLs prefixed with the `javascript:` scheme, are used as if they were a kind of navigation target but actually executes JavaScript when the browser goes to the target location.
```
Comment (hamishwillee) [inline files/en-us/web/uri/schemes/javascript/index.md:13]: Follow the model of Data urls

```suggestion
**JavaScript URLs**, URLs prefixed with the `javascript:` scheme, are used as if they were a kind of navigation target, but actually execute JavaScript when the browser goes to the target location.
```

Comment (hamishwillee) [reply files/en-us/web/uri/schemes/javascript/index.md:13]: The description is not wrong ^^^ , but it also feels a little clunky, because I think it only navigates if the javascript evaluates to something that can be navigated to. It is perhaps sufficient for an intro, but perhaps we can tidy?

```suggestion
**JavaScript URLs**, URLs prefixed with the `javascript:` scheme, are used in navigation targets, but execute JavaScript before the browser attempts to navigate.
```

Comment (Josh-Cena) [reply files/en-us/web/uri/schemes/javascript/index.md:13]: I started with this version, but "JavaScript URL" is sort of ambiguous, and the spec also uses "`javascript:` URL". I'm more inclined to rename "Data URLs" to "data: URLs" but I'm on the fence for that because "Data URL" is less ambiguous.

File: files/en-us/web/uri/schemes/javascript/index.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+## Syntax
+
+`javascript:` URLs start with the `javascript:` scheme and are followed by JavaScript code. The code will be parsed as a script.
```
Comment (hamishwillee) [inline files/en-us/web/uri/schemes/javascript/index.md:17]:
```suggestion
**JavaScript URLs** start with the `javascript:` scheme and are followed by JavaScript code. The code will be parsed as a script.
```

File: files/en-us/web/uri/schemes/javascript/index.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+```url
+javascript:<script>
+```
+
+## Description
+
+`javascript:` URLs can be used anywhere a URL is a navigation target. This includes, but is not limited to:
```
Comment (hamishwillee) [inline files/en-us/web/uri/schemes/javascript/index.md:25]: OK to use the code form here, because we're not repeating ourselves after.

File: files/en-us/web/uri/schemes/javascript/index.md (hunk 2 above)
Comment (Josh-Cena) [reply files/en-us/web/uri/schemes/javascript/index.md:13]: "JavaScript URL" is a fake navigation target. It is part of the navigation algorithm but it's possible that no "navigation" happens at all, unless the script evaluates to a string.

File: files/en-us/web/uri/schemes/javascript/index.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+When a browser attempts to navigate to such a location, it parses and executes the script body. The script may have a _completion value_ (not a return value), which is the same value if the script were executed with [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval). If the last statement is an [expression](/en-US/docs/Web/JavaScript/Reference/Statements/Expression_statement), the completion value is the value of that expression. If this completion value is a string, that string is treated as an HTML document and the browser navigates to a new page with that content, using the same URL as the current page. If the completion value is not a string, the browser only executes the code and does not navigate. Therefore, it's often recommended that if the script ends with a function call like `javascript:foo()`, you should prefix it with {{jsxref("Operators/void", "void")}} to prevent accidental navigation if the function happens to return a string.
```
Comment (hamishwillee) [inline files/en-us/web/uri/schemes/javascript/index.md:35]: not wrong, but I'd separate into two paragraphs for readability.

```suggestion
When a browser attempts to navigate to such a location, it parses and executes the script body. The script may have a _completion value_ (not a return value), which is the same value if the script were executed with [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval).

If the last statement is an [expression](/en-US/docs/Web/JavaScript/Reference/Statements/Expression_statement), the completion value is the value of that expression. If this completion value is a string, that string is treated as an HTML document and the browser navigates to a new page with that content, using the same URL as the current page. If the completion value is not a string, the browser only executes the code and does not navigate. Therefore, it's often recommended that if the script ends with a function call like `javascript:foo()`, you should prefix it with {{jsxref("Operators/void", "void")}} to prevent accidental navigation if the function happens to return a string.
```

File: files/en-us/web/uri/schemes/javascript/index.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+In this example, the `action` attribute of a `<form>` element is set to a `javascript:` URL that alerts a message when submitted:
+
+```html example-bad
+<form action="javascript:alert(myInput.value)">
+  <input id="myInput" />
+  <input type="submit" value="Submit" />
+</form>
+```
+
+Instead of doing this, consider listening for the form's `submit` event and handling it with JavaScript:
```
Comment (hamishwillee) [inline files/en-us/web/uri/schemes/javascript/index.md:83]: For all cases of ""Instead of doing this, consider" - Why?

Perhaps word as "Generally it is better to Xxxx, because YYYY". In some cases it it obvious, but even so worth spelling out.

File: files/en-us/web/uri/schemes/javascript/index.md (hunk 2 above)
Comment (hamishwillee) [reply files/en-us/web/uri/schemes/javascript/index.md:13]: The rendering is ugly and hard to parse:

image

It isn't ambiguous because you're providing the context in this page. If you're "randomly" referring to JavaScript URLs then sure - elsewhere we should refer to `javascript:` URLs.

Comment (hamishwillee) [reply files/en-us/web/uri/schemes/javascript/index.md:13]: But that's my opinion.

Comment (hamishwillee) [reply files/en-us/web/uri/schemes/javascript/index.md:13]: Maybe

```suggestion
**JavaScript URLs**, URLs prefixed with the `javascript:` scheme, are used as fake navigation targets that execute JavaScript when the browser attempts to navigate.
If the URL evaluates to a string, it is treated as HTML and rendered by the browser.
```

**Comment by hamishwillee** [comment]
In any case, this is excellent.

File: files/en-us/web/uri/schemes/javascript/index.md (hunk 1 above)
Comment (hamishwillee) [reply files/en-us/web/uri/schemes/javascript/index.md:11]: Almost exactly that. But easiest to link to a definition.

```suggestion
> Using `javascript:` URLs on the web is discouraged as it may lead to execution of arbitrary code, similar to the ramifications of using [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval). It may also reduce {{glossary("accessibility")}} because it deviates from normal link behavior.
```

File: files/en-us/web/uri/schemes/javascript/index.md (hunk 6 above)
Comment (Josh-Cena) [reply files/en-us/web/uri/schemes/javascript/index.md:83]: We are showing all ways you may use `javascript:` and suggesting alternatives, because as said in the intro there's probably no reason you should use `javascript:`, in the same way that you should never use inline HTML event handlers.

File: files/en-us/web/uri/schemes/javascript/index.md (hunk 5 above)
Comment (Josh-Cena) [reply files/en-us/web/uri/schemes/javascript/index.md]: FYI, I've reverted this change. The paragraph flows like: "the browser runs the code, the code may have a completion value, the completion value is determined by blah blah, the completion value can do blah blah". So it's hard to find a natural place to break the paragraph.

Comment (hamishwillee) [reply files/en-us/web/uri/schemes/javascript/index.md]: True. I thought this was "natural enough". Not a hill I'd die on though.

File: files/en-us/web/uri/schemes/javascript/index.md (hunk 6 above)
Comment (hamishwillee) [reply files/en-us/web/uri/schemes/javascript/index.md:84]: Fair enough. I guess I just prefer the previous cases where you were explicit - such as "This is a bad practice because the link is actually not a hyperlink. Consider making it a button instead:"

**Review by hamishwillee — APPROVED** [review]
Looks good to me. Remaining suggestions were just that. Thank you.
